python/pyspark/sql/types.py

```diff
@@ -2268,12 +2268,40 @@ def convert(self, obj: "np.generic", gateway_client: GatewayClient) -> Any:
         return obj.item()
 
 
+class NumpyArrayConverter:
+    def can_convert(self, obj: Any) -> bool:
+        return has_numpy and isinstance(obj, np.ndarray)
+
+    def convert(self, obj: "np.ndarray", gateway_client: GatewayClient) -> JavaObject:
+        from pyspark import SparkContext
+
+        gateway = SparkContext._gateway
+        assert gateway is not None
+
+        plist = obj.tolist()
+        # np.array([]).dtype is dtype('float64') so set float for empty plist
+        ptpe = type(plist[0]) if len(plist) > 0 else float
+        tpe_dict = {
+            int: gateway.jvm.int,
```

Comment:
Shouldn't we map this type from NumPy dtype?

Reply:
Since `plist = obj.tolist()`, `plist` is a list of Python scalars, see https://numpy.org/doc/stable/reference/generated/numpy.ndarray.tolist.html.

So `tpe_dict` maps Python types to Java type.

That's consistent with `NumpyArrayConverter.convert` which calls `obj.item()`, see https://numpy.org/doc/stable/reference/generated/numpy.ndarray.item.html.

Reply:
Let me know if there is a better approach :)

Reply:
Hm, unlike `obj.item` in which we have to pass Python primitive type; thus, resulting that JVM side type precision cannot be specified, here we can have more correct size in the JVM array.

I think it's better to have the correct type in the element ... Ideally we should make `obj.item` respect the numpy dtype too..

Reply:
and I believe we already have the type mapping defined in pandas API on Spark somewhere IIRC

Reply:
Makes sense!

One limitation is `np.dtype("int8")` cann't be mapped to `gateway.jvm.byte`, create `jarr` accordingly and then do the per-element assignment.

`TypeError: 'bytes' object does not support item assignment` is caused in `jarr[i] = plist[i]`.

So both `int8` and `int16` are mapped to `gateway.jvm.short`.